### ISSUE TITLE
Unbearably slow in my code base: https://github.com/scverse/scanpy

### ISSUE THREAD

**Issue by rchiodo** [microsoft/pyright#5474]
Still unbearably slow in my code base: https://github.com/scverse/scanpy

Saving files takes multiple seconds. VS Code waits and eventually shows a popup “Getting code actions from ''Python"”.

When setting `"python.languageServer": "None",` saving works as expected.

Why will VS Code wait for Pylance to save things? I have only `source.fixAll.ruff` activated, so the Python extension shouldn’t be consulted when saving a file at all.

```json
{
    "[python]": {
        "editor.formatOnSave": true,
        "editor.defaultFormatter": "ms-python.black-formatter",
        "editor.codeActionsOnSave": {
            "source.fixAll.ruff": true,
        },
    },
    "python.testing.pytestArgs": ["scanpy/tests"],
    "python.testing.unittestEnabled": false,
    "python.testing.pytestEnabled": true,
    "python.languageServer": "Pylance",
}
```

_Originally posted by @flying-sheep in https://github.com/microsoft/pylance-release/issues/4263#issuecomment-1623513714_

**Comment by rchiodo** [comment]
@Erictraut, the repro steps are in this comment here:
https://github.com/microsoft/pyright/issues/5474#issuecomment-1631605563

Same repro applies to show it taking a long time in Pyright too.

**Comment by erictraut** [comment]
This was addressed in pyright 1.1.317, which added more aggressive caching for protocol type matching. It is included in this week's "insiders release" of pylance and will be in next week's production release.

It's related to [this issue](https://github.com/microsoft/pyright/issues/5260). You can see the [measured speedup](https://github.com/microsoft/pyright/discussions/5496) in various code bases that use numpy.
